code/globalincs/pool.h

```diff
@@ -0,0 +1,404 @@
+#pragma once
+
+#include "globalincs/pstypes.h"
+#include <tl/optional.hpp>
+#include <csignal>
+//#include <vector>
+
+using  tl::optional;
+struct pool_index{
+	private:
+	int index;
+	int generation;
+	bool is_null = true;
+	public:
+	pool_index(){ nullify();};
+	pool_index(int i,int g,bool b = false){
+		if(b){ nullify();
+		}
+		else{
+			set(i,g);
+		}
+	};
+	void set(int i,int g){
+		is_null = false;
+		index = i;
+		generation = g;
+	}
+	void nullify()
+	{
+		index = -1;
+		generation = -1;
+		is_null=true;
+	};
+	bool operator==(pool_index &rhs) const {
+		if(is_null && rhs.is_null)
+			return true;
+		if(is_null || rhs.is_null)
+			return false;
+		return (index == rhs.index && generation == rhs.generation);
+	};
+	pool_index& operator=(pool_index const&rhs) {
+		if(rhs.is_null)
+			nullify();
+		else
+			set(rhs.index, rhs.generation);
+		return *this;
+	};
+	int i() const { return index;};
+	int g() const { return generation;};
+	bool null() const { return is_null;};
+	bool has_value() const { return !is_null;};
+};
+
+
+
+template  <typename T>
+struct gEntry {
+	int generation;
+	optional<T> stored;
+
+};
```

Comment:
gEntry looks to be purely internal to the pool, so it should probably be an internal subclass of pool. That way, since you also always have the same T as the pool itself, you don't need it to be a template.

Reply:
You're right, it should probably even be a private subclass.